pycalphad/io/cs_dat.py

```diff
@@ -0,0 +1,1152 @@
+    em = parse_endmember(toks, num_pure_elements + 1, num_gibbs_coeffs)
+    return EndmemberAqueous(em.species_name, em.gibbs_eq_type, em.stoichiometry_pure_elements[1:], em.intervals, em.stoichiometry_pure_elements[0])
+
+
+def parse_endmember_subq(toks: TokenParser, num_pure_elements, num_gibbs_coeffs, zeta=None):
+    em = parse_endmember(toks, num_pure_elements, num_gibbs_coeffs)
+    # The first two entries are the stoichiometry of the pair (the cation and anion). It's unclear to what the last three are for.
+    stoichiometry_quadruplet = toks.parseN(5, float)
+    if zeta is None:
+        # This is SUBQ we need to parse it. If zeta is passed, that means we're in SUBG mode
+        zeta = toks.parse(float)
+    return SUBQPair(em.species_name, em.gibbs_eq_type, em.stoichiometry_pure_elements, em.intervals, stoichiometry_quadruplet, zeta)
+
+
+def parse_quadruplet(toks):
+    quad_idx = toks.parseN(4, int)
+    quad_coords = toks.parseN(4, float)
+    return SUBQQuadrupletCoordinations(quad_idx, quad_coords)
+
+
+def parse_subq_excess(toks, mixing_type, num_excess_coeffs):
+    mixing_code = toks.parse(str)
+    mixing_const = toks.parseN(4, int)
+    mixing_exponents = toks.parseN(4, int)
+    junk = toks.parseN(12, float)
+    additional_cation_mixing_const = toks.parse(int)
+    additional_anion_mixing_exponent = toks.parse(int)
+    excess_coeffs = toks.parseN(num_excess_coeffs, float)
+    return SUBQExcessQuadruplet(mixing_type, mixing_code, mixing_const, mixing_exponents, junk, additional_cation_mixing_const, additional_anion_mixing_exponent, excess_coeffs)
+
+
+def parse_phase_subq(toks, phase_name, phase_type, num_pure_elements, num_gibbs_coeffs, num_excess_coeffs):
+    if phase_type == 'SUBG':
+        zeta = toks.parse(float)
+    else:
+        zeta = None
+    num_pairs = toks.parse(int)
+    num_quadruplets = toks.parse(int)
+    endmembers = [parse_endmember_subq(toks, num_pure_elements, num_gibbs_coeffs, zeta=zeta) for _ in range(num_pairs)]
+    num_subl_1_const = toks.parse(int)
+    num_subl_2_const = toks.parse(int)
+    subl_1_const = toks.parseN(num_subl_1_const, str)
+    subl_2_const = toks.parseN(num_subl_2_const, str)
+    subl_1_charges = toks.parseN(num_subl_1_const, float)
+    subl_1_chemical_groups = toks.parseN(num_subl_1_const, int)
+    subl_2_charges = toks.parseN(num_subl_2_const, float)
+    subl_2_chemical_groups = toks.parseN(num_subl_2_const, int)
+    subl_1_pair_idx = toks.parseN(num_subl_1_const*num_subl_2_const, int)
+    subl_2_pair_idx = toks.parseN(num_subl_1_const*num_subl_2_const, int)
+    subl_const_idx_pairs = [(s1i, s2i) for s1i, s2i in zip(subl_1_pair_idx, subl_2_pair_idx)]
+    quadruplets = [parse_quadruplet(toks) for _ in range(num_quadruplets)]
+    excess_parameters = []
+    while True:
+        mixing_type = toks.parse(int)
+        if mixing_type == 0:
+            break
+        elif mixing_type == -9:
+            # some garbage, like 1 2 3K 1 2K 1 3K 2 3 6, 90 of them
```

Comment:
Can you explain this code path? Tests don't hit it. Is this another TODO for the future?

Reply:
I have no idea what these mean and wrote this code really early in the parser implementation. The meaning isn't documented in the ChemApp documentation that I can tell. 

I have a database for Fe-Ti-V-O that ends a `SUBQ` block with 

```
   3
 G   1   4   6   6   0   1   0   0
 0.00000000       1.00 0.00000000       1.00 0.00000000       1.00
 0.00000000       0.00 0.00000000       0.00 0.00000000       0.00
   0   0 0.00000000     0.00000000     0.00000000     0.00000000
 0.00000000     0.00000000
  -9
    1    2    3K        1    2K        1    3K        2    3    6
    1    2    4K        1    2K        1    4K        2    4    6
    1    2    5K        1    2K        1    5K        2    5    6
    1    3    4K        1    3K        1    4K        3    4    6
    1    3    5K        1    3K        1    5K        3    5    6
    1    4    5K        1    4K        1    5K        4    5    6
    2    3    4K        2    3K        2    4K        3    4    6
    2    4    5K        2    4K        2    5K        4    5    6
    3    4    5K        3    4K        3    5K        4    5    6
```

The database from Starykh et al. Calphad 74 (2021) (https://doi.org/10.1016/j.calphad.2021.102312) ends a `SUBQ` block with:

```
 G   1   2   5   5   0   2   0   0
     0.00000000   1.00     0.00000000   1.00     0.00000000   1.00
     0.00000000   0.00     0.00000000   0.00     0.00000000   0.00
   0   0     0.00000000     0.00000000     0.00000000     0.00000000
     0.00000000     0.00000000
  -1
    3    1    4T3       3    1K        3    4T4       1    4    5
```

So at least I think we can infer that there's some special data here and `-N` means that there are `N` instances/lines worth of these special data.

Maybe @jorgepazsoldanpalma or @maxposchmann know what these mean?

Reply:
Those lines override the interpolation schemes selections that are normally determined by the chemical groups. We still haven't implemented this in Thermochimica, because it is a pain and not often used.

I agree with this:

    So at least I think we can infer that there's some special data here and -N means that there are N instances/lines worth of these special data.

The rest of the scheme I am not 100% sure about. I have some guesses though. For example, consider
`1    2    3K        1    2K        1    3K        2    3    6`
which I read as "for the 1-2-3 ternary, treat 1-2 as Kohler, 1-3 as Kohler, and 2-3 as Kohler". The 6 at then end indicates the constituent on the opposite sublattice.

Or in a case that uses Toop
`3    1    4T3       3    1K        3    4T4       1    4    5`
I think this means "for the 3-1-4 ternary, treat 3-1 as Toop with 3 as Toop component, 3-4 as Kohler, and 1-4 as Toop with 4 as Toop component", where now 5 is the anion constituent.

I guess this is necessary if the desired selection of interpolation schemes does not follow the chemical group rules.

Reply:
Thanks, @maxposchmann. I am following along and that would seem to make sense. For the SUBQ cases where there is mixing in the cation sublattice and there is a summation over the anions (e.g. your 2021 paper Eq. 22), you are using the these relationships for the anion in the interaction parameter - i.e. anion `k`, not `x` or `y` (using your notation) - right?

@richardotis: I already pushed some changes so that we can parse successfully both cases, but we currently just throw away this metadata.

I think I have some idea on how to handle this in pycalphad. In the `ModelMQMQA._chemical_group_filter`, we could have an option argument (or some state stored) that contains "overrides" that use the specified relationships rather than the chemical group relationships with the behavior here. The painful part is that the filter function now needs to know what anions is being considered. Since we have some databases, but Thermochimica does not yet have this feature, we probably need FactSage to test and compare against if we want the true answer. Some options:

1. Implement in this PR and have someone check some test calculations against FactSage to add to the test suite
2. Defer implementing and raise an error if these overrides are parsed. The rationale for raising an error is that incorrect results should be impossible
3. Defer implementing and issue a warning if these overrides are parsed. The rationale for a warning is that these overrides only impact the calculations and extrapolation on some specific subsystems, so a user may want to do calculations in a subsystem where these overrides do not apply and our calculations will be correct.

Reply:
Defer and warn is fine. I think our DAT support is going to have to be considered preliminary at this point, given the lack of a real specification, but it's a whole lot better than nothing.

Reply:
Sounds good. I did some of the legwork to parse the overrides into a dictionary form, enough to be able to warn users which sets of elements may be incorrect in 0689134a5d8e1b278502aabe3ee9e9312ed3bfd2. For example:

- The string `1 2 3K 1 2K 1 3K 2 3 6` gives a warning:

```python
<snip>/pycalphad/pycalphad/io/cs_dat.py:721: UserWarning: Phase SLAG-LIQ overrides the ternary extrapolation models for the system ['CU1+', 'SI', 'SN4+', 'O']. Use caution as extrapolated energies may be incorrect.
```

- The string `3 1 4T3 3 1K 3 4T4 1 4 5` gives a warning:

```python
<snip>/pycalphad/pycalphad/io/cs_dat.py:721: UserWarning: Phase SLAGBSOLN overrides the ternary extrapolation models for the system ['FE[3+]', 'FE[2+]', 'V2O3', 'O']. Use caution as extrapolated energies may be incorrect.
```

Reply:
I'm not sure I followed your question correctly. I'll set up an example that matches my understanding of what you asked, and you can let me know if I missed the mark.

Let's say we have cations 1, 2, 3, and anions 4, 5. I won't assign then chemical groups, but suppose we overrode them for {1,2,3,5} but not {1,2,3,4}. Your question is then, how do we compute the result of equation 22 for say {1,2/4,4}?

If so, I agree that this is not well-defined in my notation. I think your suggested approach of using the mixing as defined for cation k is likely correct. After all, given the deltas, we only accumulate into the sum for x or y equal to k.